### DISCUSSION TITLE
Blur hash (base64) for images

### DISCUSSION THREAD

**Discussion by Francismori7** [Shopify/storefront-api-feedback#75] [Ideas & Requests]
Hey!

Just wondering if perhaps the storefront API could make available a blur hash within images. I was planning on implementing something like that on Shopify Hydrogen, but since that would increase page load, that seems counterintuitive.

If however the storefront API itself made it available, we could use the hash to display a blurred image (see Plaiceholder or Next/Image) until the requested image is fully loaded on the client.

Since the Shopify CDN already plays a role into image optimization, I'm not sure it would be a big issue implementing a basic blurhash on it.

Thanks!

**Comment by frehner** [comment]
Blurhashes would be a _great_ addition, and would be a nice improvement that we could add to Hydrogen's `<Image/>` component!

**Comment by jobveldhuis** [comment]
Seconding this. I'm currently running into places where I really want to give my user some visual feedback while lazy loading big images and incorporated blurhash would make my life so much better.

**Comment by fcisio** [comment]
Has there been any progress on this? I would also really like to see this feature.

**Comment by EricWVGG** [comment]
So it’s already there, just not documented. There's a fair amount of info on rendering them in React Native, for some reason, but next to none for React…

You'll need to [install the thumbhash lib](https://www.npmjs.com/package/thumbhash?activeTab=versions) `npm install thumbhash`

If you go into your Graphiql explorer, you'll see that along with `width, height, url`, there’s a property called `thumbhash`. Add that to any GraphQL query that you want to use this with.

```typescript
// thumbhash.ts
import { thumbHashToDataURL } from 'thumbhash'

// this converts the thumbhash from a base64-encoded string to a readable integer array.
export function base64ToUint8Array(base64: string): Uint8Array {
  const decodedString = atob(base64)
  const arrayBuffer = new Uint8Array(decodedString.length)
  for (let i = 0; i < decodedString.length; i++) {
    arrayBuffer[i] = decodedString.charCodeAt(i)
  }
  return arrayBuffer
}

// this is a convenience function
export const thumbhash = (hash?: string | null) => {
  if(!hash) return undefined
  return thumbHashToDataURL(base64ToUint8Array(hash))
}
```

```typescript
// ProductImage.tsx
import type { ProductVariantFragment } from 'storefrontapi.generated'
import {thumbhash} from 'thumbhash.ts'

export function ProductImage({
  image,
  className,
}: {
  image: ProductVariantFragment['image']
  className?: string
}) {
  const dataUrl = thumbhash(image?.thumbhash)

  return (
    <div className='product-image image-placeholder' style={{ backgroundImage: `url(${dataUrl})` }}>
      {image && (
        <Image
          alt={image.altText || 'Product Image'}
          aspectRatio='1/1'
          data={image}
          key={image.id}
          sizes='(min-width: 45em) 50vw, 100vw'
        />
      )}
    </div>
  )
}
```

```css
// some-stylesheet.css

.image-placeholder {
  background-size: cover;
  background-repeat: no-repeat;
}
```
